internal/executor/cache/cache.go

```diff
@@ -0,0 +1,78 @@
+package cache
+
+import (
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+var (
+	ErrFailedToInitialize = errors.New("cache initialization failed")
+	ErrBlobNotFound       = errors.New("blob with the specified key not found")
+	ErrInternal           = errors.New("internal cache error")
+)
+
+type Cache struct {
+	namespaceDir string
+}
+
+func New(dir string, namespace string) (*Cache, error) {
+	if dir == "" {
+		userCacheDir, err := os.UserCacheDir()
+		if err != nil {
+			return nil, fmt.Errorf("%w: %v", ErrFailedToInitialize, err)
+		}
+		dir = userCacheDir
+	}
+
+	namespaceDir := filepath.Join(dir, "cirrus", namespace)
+
+	// Create a base directory, ignoring ErrExist since it may already be created
+	// by a previous or parallel invocation of the CLI
+	if err := os.MkdirAll(namespaceDir, 0700); err != nil {
+		if !os.IsExist(err) {
+			return nil, fmt.Errorf("%w: %v", ErrFailedToInitialize, err)
+		}
+	}
+
+	return &Cache{
+		namespaceDir: namespaceDir,
+	}, nil
+}
+
+func (c *Cache) Get(key string) (*os.File, error) {
+	file, err := os.OpenFile(c.blobPath(key), os.O_RDONLY, 0)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrBlobNotFound
+		}
+		return file, fmt.Errorf("%w: %v", ErrInternal, err)
+	}
+
+	return file, nil
+}
+
+func (c *Cache) Put(key string) (*PutOperation, error) {
+	tmpBlobFile, err := ioutil.TempFile(c.namespaceDir, ".temporary-blob-")
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", ErrInternal, err)
+	}
+
+	return &PutOperation{
+		tmpBlobFile:   tmpBlobFile,
+		finalBlobPath: c.blobPath(key),
+	}, nil
+}
+
+func (c *Cache) blobPath(key string) string {
+	// Sanitize user-controlled data by hashing it
```

Comment:
Might not be necessary since cache keys are always in this form `<cache name>-<sha256>`. Plus will be easier for users to find the blobs for manual inspection.

Reply:
It's technically possible for the arbitrary code that runs in the task's environment to impersonate the agent and connect to the CLI over new (or existing) RPC connection.

So this depends on whether we trust this environment or not.

Reply:
Alternatively, we can validate the key using a simple regular expression.

Reply:
Yeah, can do something like:

```suggestion
	isSafeChar := func(c rune) bool { 
		if '0' <= c && c <= '9' { return true }
		if 'a' <= c && c <= 'z' { return true }
		if 'A' <= c && c <= 'Z' { return true }
		if c == '-' || c == '_' { return true }
		return false
	}
	isNotSafeChar := func(c rune) bool { 
		return !isSafeChar(c)
	}
	containsNonSafeChar := strings.IndexFunc(key, isNotSafeChar) > 0
	if !containsNonSafeChar {
		// only safe characters so no need for a sanitization
		return key
	}
	// Sanitize user-controlled data by hashing it
```

Reply:
See https://github.com/cirruslabs/cirrus-cli/pull/17/commits/379f4a2983affb6a68db75cbf80e38034e0ccb89.